docs/dimbl.1
.TH dimbl 1 "2014 december 02"

.SH NAME
dimbl \- Distributed Timbl
.SH SYNOPSIS
dimbl [timbl options] [\-S <num of threads>]

.SH DESCRIPTION
dimbl extends
.B timbl
with the possibility to run the classification task on multiple threads.
This is done by splitting up the Instancebase in parts that are handled in
parallel. Every test instance is tested against each partial Instancebase.
The results are merged and then the k Nearest Neighbours are calculated.

.SH NOTES
dimbl only works for the IB1 variants of timbl.

Not all timbl options are fully supported.

Documentation is lacking.

.SH OPTIONS

.B \-S
<threads>
.RS
run the server on 'threads' parallel threads.
.RE

all
.B timbl
options are documenented in
.BR timbl (1)

dimbl handles most of them in the same way as
.B timbl
except for

.BR \-i " filename"

and

.BR \-I " filename"

.B \-I
will create 'threads' Instancebase files and store their names in 'filename'
together with the name of the weighting file.

.B \-i
will use such a file to read back in 'threads' Instancebases for a classifying
task. The
.B \-S
option is ignored in that case.
.B dimbl
will use the number of files found in 'filename'.

.SH BUGS
possibly

.SH AUTHORS
Ko van der Sloot Timbl@uvt.nl

Antal van den Bosch Timbl@uvt.nl

.SH SEE ALSO
.BR timbl (1)


NEWS
0.15 - 2016-01-14
[Ko van der Sloot]
 * repository moved to GitHub
 * added Travis support
 * depends on ticuutils now.

0.14 - 2014-09-23
[Ko van der Sloot]
 * fixed make 'distcheck', store intermediate results in /tmp
 * release

0.13 lost in transition

0.12 - 2013-04-04
 * depends on ticcutils now
 * some cleanup in configure.ac

0.11 -2011-12-22
 * Added a test for 'make check'
 * use standard exit codes (EXIT_FAILURE/EXIT_SUCCES)

0.10 -2011-08-30
* CNTS ==> CLiPS
* some more decapping

0.9 - 2011-03-21

[ Ko van der Sloot]
* decapped to dimbl

0.8 - 2011-02-15

[ Antal van den Bosch ]
* m4/openmp.m4: added
* configure.ac: changed openmp check to the one in Frog

[ Joost van Baal ]
* Makefile.am: do not install ChangeLog

[ Ko van der Sloot ]
* docs/Dimbl.1: Added man page


0.7 - 2010-06-02

[ Ko van der Sloot ]
* See ChangeLog


0.4 - 2009-01-12

[ Ko van der Sloot ]
* 2008-06-09 Fixed Dimpl (openMP version)
* 2008-06-02 we use SVN now!


0.3 - 2008-04-04

[ Ko van der Sloot ]
* See ChangeLog


0.2-beta - 2008-03-01

[ Ko van der Sloot ]
* 2007-12-03 Finished packaging (for now)
* 2007-10-03 Started packaging


src/Dimbl.cxx
/*
  Copyright (c) 2006 - 2017
  CLST  - Radboud University
  ILK   - Tilburg University
  CLiPS - University of Antwerp

  This file is part of dimbl

  dimbl is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  dimbl is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/dimbl/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <fstream>
#include <vector>
#include <cstdlib>
#include <sys/time.h>
#include <omp.h>
#include <config.h>
#include "timbl/TimblAPI.h"
#include "dimbl/DimProcs.h"

using namespace std;
using namespace Timbl;
using namespace TiCC;

//#define DEBUG 1

class mp_worker: public worker {
public:
  bool Init( TimblAPI *, const string&, const string&, int );
  bool writeTree( const string& );
  bool readTree( const TiCC::CL_Options&, const string&, const string&, int );
  bool Execute( const string & );
};

bool mp_worker::Init( TimblAPI *pnt,
		      const string& fileName,
		      const string& wfName, int i ){
  exp = pnt;
  cnt = i;
#if DEBUG > 0
#pragma omp critical
  cerr << "starting worker[" << cnt << "], thread = "
       << omp_get_thread_num()+1 << "/" << omp_get_num_threads() << endl;
#endif
  bool ok = exp->Learn( fileName );
  if ( !ok ){
    cerr << "learning failed for " << fileName << endl;
    cerr << "stopped" << endl;
    return false;
  }
  if ( !wfName.empty() )
    ok = exp->GetWeights( wfName, exp->CurrentWeighting() );
  if ( !ok ){
    cerr << "reading weights from " << wfName << " failed " << endl;
    cerr << "stopped" << endl;
    return false;
  }
  return true;
}

bool mp_worker::Execute( const string& line ){
  result = 0;
  if ( !line.empty() ){
#if DEBUG > 0
#pragma omp critical
    cerr << "exp-" << cnt << " thread = "
	 << omp_get_thread_num()+1 << "/" << omp_get_num_threads()
 	 << " classify '" << line << "'" << endl;
#endif
    result = exp->classifyNS( line );
    result->setShowDistance( true );
    result->setShowDistribution( true );
  }
  return true;
}

bool mp_worker::readTree( const TiCC::CL_Options& opts,
			  const string& name,
			  const string& wfName,
			  int i ){
  bool ok = false;
  exp = new TimblAPI( opts );
  cnt = i;
#if DEBUG > 0
#pragma omp critical
  cerr << "starting worker[" << cnt << "], thread = "
       << omp_get_thread_num()+1 << "/" << omp_get_num_threads() << endl;
#endif
  if ( exp->GetInstanceBase( name ) ){
    ok = exp->GetWeights( wfName, exp->CurrentWeighting() );
    if ( !ok ){
      cerr << "reading weights from " << wfName << " failed " << endl;
      cerr << "stopped" << endl;
    }
  }
  time_stamp( cout, "initialize a tree from " + name );
  return ok;
}

bool mp_worker::writeTree( const string& name ){
  return exp->WriteInstanceBase( name );
}

template <>
bool experiment<mp_worker>::createWorkers(){
  bool ok = true;
  if ( size < 1 )
    size = omp_get_max_threads();
  else {
    omp_set_num_threads( size );
    int mt = omp_get_max_threads();
    if ( mt < size ){
      cout << "couldn't get more then " << mt << " threads (requested was "
	   << size << " )" << endl;
      size = mt;
    }
  }
  children.resize( size );
  split( config.trainFileName, size, config.tmpdir );
  time_stamp( cout, "   start training "+ toString(size) + " children:" );
  int i;
#pragma omp parallel private( i )
  {
    for( i=1; i <= size; ++i ){
#pragma omp single nowait
      {
	string iname = config.tmpdir
	  + TiCC::basename(config.trainFileName) + "-" + toString( i );
#if DEBUG > 0
	cerr << "start " << iname << endl;
#endif
	TimblAPI *child = new TimblAPI( *train );
	children[i-1].Init( child, iname, config.weightsFileName, i );
	cout << "child " << i << ":";
	children[i-1].exp->ShowIBInfo( cout );
      }
    }
  }
  time_stamp( cout, "finished training children:" );
  return ok;
}

template <>
bool experiment<mp_worker>::createWorkerFiles(){
  if ( size < 1 )
    size = omp_get_max_threads();
  else {
    omp_set_num_threads( size );
    int mt = omp_get_max_threads();
    if ( mt < size ){
      cout << "couldn't get more then " << mt << " threads (requested was "
	   << size << " )" << endl;
      size = mt;
    }
  }
  children.resize( size );
  split( config.trainFileName, size, config.tmpdir );
  time_stamp( cout, "   start training "+ toString(size) + " children:" );
  ofstream os( config.treeOutFileName );
  int i;
  vector<bool> ok(size,true);
#pragma omp parallel private( i )
  {
    for( i=1; i <= size; ++i ){
#pragma omp single nowait
      {
	string iname = config.tmpdir
	  + TiCC::basename(config.trainFileName) + "-" + toString( i );
#if DEBUG > 0
	cerr << "start " << iname << endl;
#endif
	TimblAPI *child = new TimblAPI( *train );
	children[i-1].Init( child, iname, config.weightsFileName, i );
	string oname = config.tmpdir
	  + TiCC::basename(config.trainFileName) + "-" + toString( i ) + ".tree";
	os << oname << endl;
#if DEBUG > 0
	cerr << "start saving in" << oname << endl;
#endif
	if ( !children[i-1].writeTree( oname ) )
	  ok[i-1] = false;
	else {
#pragma omp critical
	  {
	    cout << "child " << i << ":";
	    children[i-1].exp->ShowIBInfo( cout );
	  }
	}
      }
    }
  }
  bool result = true;
  for( i=0; i < size; ++i ){
    result &= ok[i];
  }
  if ( result ){
    time_stamp( cout, "finished saving trees" );
    string wF =  config.weightsFileName;
    if ( wF.empty() )
      wF = config.treeOutFileName + ".wghts";
    result = train->SaveWeights( wF );
    if ( !result ){
      cerr << "saving weights in " << wF << " failed " << endl;
      cerr << "stopped" << endl;
    }
    else {
      cerr << "weights saved in " << wF << endl;
      os << wF << endl;
      time_stamp( cout, "filenames stored in: " + config.treeOutFileName );
    }
  }
  else
    time_stamp( cout, "failed saving trees" );
  return result;
}

template <>
bool experiment<mp_worker>::createWorkersFromFile( const TiCC::CL_Options& opts ){
  ifstream is( config.treeInFileName );
  if ( !is ){
    cerr << "unable to read file: " << config.treeInFileName << endl;
    return false;
  }
  vector<string> filenames;
  string line;
  while ( getline( is, line ) ){
    ifstream tmp( line );
    if ( !tmp ){
      cerr << "unable to read file: "<< line << endl;
      return false;
    }
    filenames.push_back( line );
  }
  size = filenames.size() - 1;
  string wFile;
  if ( size == 0 ){
    cerr << "no usefull info in" <<  config.treeInFileName << endl;
    return false;
  }
  else {
    wFile = filenames[size];
    ifstream tmp( wFile );
    if ( !tmp ){
      cerr << "unable to read weightsfile: " << wFile << endl;
      return false;
    }
    else
      cerr << "using weightsfile: " << wFile << endl;
  }
  omp_set_num_threads( size );
  int mt = omp_get_max_threads();
  if ( mt < size ){
    cerr << "couldn't get more then " << mt << " threads (requested was "
	 << size << " )" << endl;
    return false;
  }
  children.resize( size );
  time_stamp( cout, "   start reading input for "+ toString(size) + " children:" );
  int i;
  vector<bool> ok(size,true);
#pragma omp parallel private( i )
  {
    for( i=1; i <= size; ++i ){
#pragma omp single nowait
      {
	string iname = filenames[i-1];
#if DEBUG > 0
	cerr << "start " << iname << endl;
#endif
	if ( !children[i-1].readTree( opts, iname, wFile, i ))
	  ok[i-1] = false;
      }
    }
  }
  bool result = true;
  for( i=0; i < size; ++i ){
    result &= ok[i];
  }
  if ( result ){
    time_stamp( cout, "finished initializing children:" );
  }
  return result;
}

void usage(){
  cerr << "usage:" << "\tdimbl -S threads [Timbl Options]" << endl
       << "\twhere 'threads' is a number specifying the number of" << endl
       << "\tthreads that is used for training" << endl
       << "\t'Timbl Options' specifies the Timbl Options to use." << endl
       << "\tNote that not all Timbl options are working yet" << endl
       << "\tmost usefull are: -f, -t, -i " << endl;
}

int main(int argc, char *argv[]) {
  cerr << "dimbl " << VERSION << " (c) CLST/ILK 1998 - 2017" << endl;
  cerr << "Centre for Language and Speech Technology, Radboud University" << endl;
  cerr << "Induction of Linguistic Knowledge Research Group, Tilburg University" << endl;
  cerr << "based on [" << Timbl::VersionName() << "]" << endl;
  if ( argc < 2 ){
    usage();
    return EXIT_SUCCESS;
  }
  try {
    TiCC::CL_Options opts;
    opts.init( argc, argv ); // we don't check the arguments for now
    string value;
    if ( opts.is_present( 'h' ) ){
      usage();
      return EXIT_SUCCESS;
    }
    if ( opts.is_present( 'V' ) || opts.is_present( "version" ) ){
      return EXIT_SUCCESS;
    }
    experiment<mp_worker> theExp( opts );
    if ( !theExp.config.treeInFileName.empty() ){
      if ( !theExp.config.treeOutFileName.empty() ){
	cerr << "both -i and -I specified! That is impossible to handle"
	     << endl;
	return EXIT_FAILURE;
      }
      if ( !theExp.createWorkersFromFile( opts ) )
	return EXIT_FAILURE;
    }
    else {
      theExp.Train( opts );
      if ( !theExp.config.treeOutFileName.empty() ){
	theExp.createWorkerFiles();
	cout << "done." << endl;
	return EXIT_SUCCESS;
      }
      else
	if ( !theExp.createWorkers() )
	  return EXIT_FAILURE;
    }
    int lineCount = 0;
    time_t startTime;
    time(&startTime);
    timeval Start;
    gettimeofday( &Start, 0 );
    string line;
    theExp.initStatistics();
    time_stamp( cout, "start testing '" + theExp.config.testFileName + "'" );
    while ( getline( theExp.config.inp, line ) ){
      ++lineCount;
      if ( line.empty() )
	continue;
      int i;
#pragma omp parallel for private(i)
      for ( i=0; i < theExp.size; ++i ){
	theExp.children[i].Execute( line );
      }
      neighborSet result = theExp.Finalize();
      theExp.showResult( result, line );
      theExp.showProgress( cout, lineCount, startTime );
    }
    time_stamp( cout, "Ready:  ", lineCount );
    theExp.showStatistics( cout );
    show_speed_summary( cout, lineCount, Start );
    cout << "results in " << theExp.config.outFileName << endl;
  }
  catch ( exception& e ){
    cerr << "terminated because of " << e.what() << endl;
    return EXIT_FAILURE;
  }
  return EXIT_SUCCESS;
}
